packages/teams-js/src/internal/videoUtils.ts

```diff
@@ -78,36 +108,142 @@ function createProcessedStreamGenerator(
+    const timestamp = originalFrame.timestamp;
+    if (timestamp !== null) {
+      try {
+        const { videoFrame, metadata: { audioInferenceResult } = {} } = await this.extractVideoFrameAndMetadata(
+          originalFrame,
+        );
+        const frameProcessedByApp = await this.videoFrameHandler({ videoFrame, audioInferenceResult });
+        // the current typescript version(4.6.4) dosn't support webcodecs API fully, we have to do type conversion here.
+        const processedFrame = new VideoFrame(frameProcessedByApp as unknown as CanvasImageSource, {
+          // we need the timestamp to be unchanged from the oirginal frame, so we explicitly set it here.
+          timestamp: timestamp,
+        });
+        controller.enqueue(processedFrame);
+        videoFrame.close();
+        originalFrame.close();
+        (frameProcessedByApp as VideoFrame).close();
+      } catch (error) {
+        originalFrame.close();
+        this.notifyError(error);
+      }
+    } else {
+      this.notifyError('timestamp of the original video frame is null');
+    }
+  };
+
+  /**
+   * @hidden
+   * Extract video frame and metadata from the given texture.
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  private extractVideoFrameAndMetadata = async (
+    texture: VideoFrame,
+  ): Promise<{ videoFrame: VideoFrame; metadata: { audioInferenceResult?: Uint8Array } }> => {
+    if (inServerSideRenderingEnvironment()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    if (texture.format !== 'NV12') {
+      this.notifyError('Unsupported video frame format');
+      throw new Error('Unsupported video frame format');
+    }
+
+    /**
+     * stream id for audio inference data
+     */
+    const AUDIO_INFERENCE_RESULT_STREAM_ID = 0x31646961;
+
+    // The rectangle of pixels to copy from the texture
+    const headerRect = { x: 0, y: 0, width: texture.codedWidth, height: 2 };
+    // it's in NV12 format, but the real data is in the y plane only.
+    const headerBuffer = new ArrayBuffer((headerRect.width * headerRect.height * 3) / 2);
+    await texture.copyTo(headerBuffer, { rect: headerRect });
+    const headerDataView = new Uint32Array(headerBuffer);
+    // const [ oneTextureId, version, frameRowOffset, frameFormat, frameWidth, frameHeight, multiStreamHeaderRowOffset, multiStreamCount ] = headerDataView;
+    const metadataRect = {
+      x: 0,
+      y: headerDataView[6], // multiStreamHeaderRowOffset
+      width: texture.codedWidth,
+      height: texture.codedHeight - headerDataView[6], // multiStreamHeaderRowOffset
+    };
+    const metadataBuffer = new ArrayBuffer((metadataRect.width * metadataRect.height * 3) / 2);
+    await texture.copyTo(metadataBuffer, { rect: metadataRect });
+    const metadata = new Uint32Array(metadataBuffer);
+    let audioInferenceResult: Uint8Array | undefined;
+    for (let i = 0, index = 0; i < headerDataView[7] /* multiStreamCount */; i++) {
```

Comment:
7? How do you know how many entries there are in this array? You should have protection against index out of bounds